### DISCUSSION TITLE
Should dagger secret be its own type?

### DISCUSSION THREAD

**Discussion by fkautz** [dagger/dagger#349] [Ideas]
The current dagger input looks like this:

```sh
✗ dagger input
Manage a deployment's inputs

Usage:
  dagger input [command]

Available Commands:
  container   Add a container image as input artifact
  dir         Add a local directory as input artifact
  git         Add a git repository as input artifact
  json        Add a JSON input
  secret      Add an encrypted input secret
  text        Add a text input
  yaml        Add a YAML input
```

Should dagger secret be its own individual type? Or should secret be a property which can be applied to other types?

e.g. `dagger json --secret` or `dagger secret --type=json name encrypted-input`

Alternatively, we could just leave secret as is and treat it as a string or byte based on how it is consumed.

Also, there is an issue to be careful with here. cue strings are utf-8 which breaks if the input byte string is not strictly utf-8 compatible.

We should also consider the input source. e.g. source could be cli, stdin, file, envier, url, or other locations to be invented in the future. :)

**Comment by shykes** [comment]
I think it should be its own type, because dagger handles it in a very special way, and it is a very differentiating feature so it makes sense to highlight it. The same is true of artifacts.

I think you can say there are 3 possible types of inputs:

1. Regular values (source: any json-compatible format)
2. Encrypted secrets (source: any json-compatible format)
3. Artifacts (source: any llb-compatible data source)

**Reply by fkautz** [reply]
For point 2, what if the encrypted secret is an X509 cert or something else not json compatible? We could base64 encode it before passing it in, though that seems like an extra step that may not be necessary when cue supports bytes naturally.
